### ISSUE TITLE
Add proper right shifts

### ISSUE THREAD

**Issue by bbrk24** [bbrk24/Trilangle#41]
TL;DR: there is no (arithmetic or logical) right shift operator, and one would be useful.

---

As @bigyihsuan pointed out to me on Discord a while ago, there is no reliable way to right-shift a negative value. The closest you can get is something like this:

```
     BNG neg
     PSI #2
     DIV
     JMP next
neg: NOT
     PSI #2
     DIV
     NOT
next:
```

This branching construct is rather bulky, discouraging its use in code. As I think about it, I realize there is no proper *left* shift either: I included `EXP` to sidestep this, expecting it to be used in conjunction with `MUL` (for left shift) and `DIV` (for right shift). However, signed truncating division is not suitable for right-shifting negative values in the general case: `(-3) >> 2` should be either -1 (sign extension) or 4_194_303 (zero extension), and `-3 / 4` is 0.

There are several possible solutions:

- [![semver-major][semver-major-shield]][semver-major-label] Change `DIV` from truncating to rounding down. I believe this makes the instruction suitable for ASR, though I haven't proven it.
- [![semver-minor][semver-minor-shield]][semver-minor-label] Add a `UDV` instruction for unsigned division, suitable for LSR.
- [![semver-minor][semver-minor-shield]][semver-minor-label] Add binary instructions for `ASR` and/or `LSR`, leaving an asymmetry with left-shift which is still two instructions (`EXP`+`MUL`).
  - [![semver-minor][semver-minor-shield]][semver-minor-label] Also add `SHL` which is shorthand for `EXP`+`MUL`. Adding shorthands has some precedent (see `DP2`).
    - [![semver-major][semver-major-shield]][semver-major-label] Also deprecate/remove `EXP`.
- [![semver-minor][semver-minor-shield]][semver-minor-label] Add carryless multiplication, and a custom operation that potentially performs multiple rightshifts and ORs them the same way carryless multiplication does leftshifts. The opcodes could be `CLM` (**c**arry**l**ess **m**ultiply) and `BDO` (**b**itwise <code><strong>D</strong>IV</code>+<code><strong>O</strong>R</code>).  
  For each set bit in y, multiply/divide x by just that bit, then OR the results together. For example:
  - x=4,y=6: `CLM = 4*4|4*2 = 24`, `BDO = 4/4|4/2 = 3`
  - x=6,y=2: `CLM = 6*2 = 12`, `BDO = 6/2 = 3`
  - x=6,y=6: `CLM = 6*4|6*2 = 28`, `BDO = 6/4|6/2 = 3`
  - y=0: `CLM = 0`, `BDO = 0`
  
  If `BDO` is implemented with unsigned division, `EXP` + `CLM`/`BDO` work as symmetric logical shifts.

[semver-major-shield]: <https://img.shields.io/github/labels/bbrk24/Trilangle/semver-major> "semver-major"
[semver-major-label]: https://github.com/bbrk24/Trilangle/labels/semver-major
[semver-minor-shield]: <https://img.shields.io/github/labels/bbrk24/Trilangle/semver-minor> "semver-minor"
[semver-minor-label]: https://github.com/bbrk24/Trilangle/labels/semver-minor

**Comment by bbrk24** [comment]
After some thought, I'm leaning towards UDV or BDO. I want to keep EXP's utility, and symmetry between left and right shifts. Signed and unsigned addition/subtraction/multiplication are already the same operation, so only division needs separate instructions for the two; however, there aren't generally unsigned operations (i.e. no unsigned PTI or MOD).
